portability-gateway/src/main/java/org/dataportabilityproject/gateway/action/startjob/StartJobAction.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2018 The Data-Portability Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.dataportabilityproject.gateway.action.startjob;
+
+import com.google.api.client.util.Sleeper;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.io.BaseEncoding;
+import com.google.inject.Inject;
+import java.io.IOException;
+import java.security.PublicKey;
+import java.util.UUID;
+import org.dataportabilityproject.gateway.action.Action;
+import org.dataportabilityproject.gateway.action.ActionUtils;
+import org.dataportabilityproject.security.AsymmetricKeyGenerator;
+import org.dataportabilityproject.security.Encrypter;
+import org.dataportabilityproject.security.EncrypterFactory;
+import org.dataportabilityproject.spi.cloud.storage.JobStore;
+import org.dataportabilityproject.spi.cloud.types.JobAuthorization;
+import org.dataportabilityproject.spi.cloud.types.JobAuthorization.State;
+import org.dataportabilityproject.spi.cloud.types.PortabilityJob;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** An {@link Action} that starts a transfer job. */
+public final class StartJobAction
+    implements Action<StartJobActionRequest, StartJobActionResponse> {
+
+  private static final Logger logger = LoggerFactory.getLogger(StartJobAction.class);
+
+  private final JobStore store;
+  private final AsymmetricKeyGenerator asymmetricKeyGenerator;
+
+  @Inject
+  StartJobAction(
+      JobStore store,
+      AsymmetricKeyGenerator asymmetricKeyGenerator) {
+    this.store = store;
+    this.asymmetricKeyGenerator = asymmetricKeyGenerator;
+  }
+
+  /**
+   * Starts a job using the following flow:
+   *
+   * <li>Validate auth data is present in cookies
+   * <li>Set Job to state pending assignment
+   * <li>Wait for a worker to be assigned
+   * <li>Once worker assigned, grab worker key to encrypt auth data from cookies
+   * <li>Update job with auth data
+   */
+  @Override
+  public StartJobActionResponse handle(StartJobActionRequest request) {
+    UUID jobId = request.getJobId();
+    // Update the job to indicate to worker processes that creds are available for encryption
+    updateStateToCredsAvailable(jobId);
+
+    // Poll and block until a public key is assigned to this job
+    PortabilityJob job = pollForPublicKey(jobId);
+
+    // Update this job with
```

Comment:
... credentials encrypted with the worker's public key

Reply:
Done